blueprint/core/blueprint_ViewManager.h

```diff
@@ -0,0 +1,87 @@
+/*
+  ==============================================================================
+
+    blueprint_ViewManager.h
+    Created: 21 May 2020 15:32:00am
+
+  ==============================================================================
+*/
+
+#pragma once
+
+#include <map>
+
+#include "blueprint_View.h"
+#include "blueprint_ShadowView.h"
+
+namespace blueprint
+{
+    //TODO: Document
+    class ViewManager : public juce::ComponentListener
+    {
+    public:
+        //==============================================================================
+        // We allow registering arbitrary view types with the React context by way of
+        // a "ViewFactory" here which is a user-defined function that produces a View
+        // and a corresponding ShadowView.
+        typedef std::pair<std::unique_ptr<View>, std::unique_ptr<ShadowView>> ViewPair;
+        typedef std::function<ViewPair()> ViewFactory;
+
+        //==============================================================================
+
+        //TODO: Document
+        explicit ViewManager(View* rootView);
+        ~ViewManager();
+
+        /** Registers a new dynamic view type and its associated factory. */
+        void registerViewType(const juce::String& typeId, ViewFactory f);
+
+        /** Creates a new view instance and registers it with the view table. */
+        ViewId createViewInstance(const juce::String& viewType);
+
+        /** Creates a new text view instance and registers it with the view table. */
+        ViewId createTextViewInstance(const juce::String& value);
+
+        //TODO: Document
+        void setViewProperty(ViewId viewId, const juce::Identifier& name, const juce::var& value);
+
+        //TODO: Document
+        void setRawTextValue(ViewId viewId, const juce::String& value);
+
+        //TODO: Document
+        void addChild(ViewId parentId, ViewId childId, int index = -1);
+
+        //TODO: Document
+        void removeChild (ViewId parentId, ViewId childId);
+
+        //==============================================================================
+        // juce::ComponentListener
+        void componentMovedOrResized (juce::Component& component, bool wasMoved, bool wasResized) override;
+
+        //==============================================================================
+    private:
+        void enumerateChildViewIds (std::vector<ViewId>& ids, View* v);
+
+        /** Recursively computes the shadow tree layout on the root ShadowView, then traverses the tree
+            flushing new layout bounds to the associated view components.
+         */
+        void performRootShadowTreeLayout();
+
+        /** Returns a pointer pair to the view associated to the given id. */
+        //TODO: Private for now as only used internally but any desire for this to be made public?
```

Comment:
Good question. I think as long as it's used only internally we should keep it private. If and when we have a good case to make it public we can do it then

Reply:
Will kill the TODO in next set of changes.